docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,213 @@
+---
+title: Configuration Management
+tags: [Reference, Device Profile, Firmware Management, Software Management, Configuration Management]
+sidebar_position: 6
+description: Device profile API proposal
+---
+
+# Device profile
+
+A device profile defines any desired combination of a firmware, software and associated configurations to be installed on a device.
+Device profiles are used to get a fleet of devices into a consistent and homogenous state by having the same set of firmware,
+software and configurations installed on all of them.
+
+The `tedge-agent` handles `device_profile` operations as follows:
+
+* Declares `device_profile` support by sending the capability message to `<root>/<device-topic-id>/cmd/device_profile` topics
+* Subscribes to `<root>/<device-topic-id>/cmd/device_profile/+` to receive `device_profile` commands.
+* The `device_profile` command payload is an ordered list of modules representing any firmware, software and configuration combination.
+* The agent processes each module one by one in the same order, triggering sub-operations for the respective module type.
+* One successful installation of all the modules, the applied profile information is published to the same capability topic.
+* No rollback is performed on partial failures unless the subcommand for the failed module can rollback that single module.
+
+# Why device profile FAQ
+
+**Q1: Why do we need device profile when firmware, software and configuration management already exists**
+
+Performing and managing these operations individually for a long list of software and configuration items would be cumbersome,
+especially when performed on a large fleet of devices as each operation will have to be monitored and managed separately.
+Grouping them together into a single operation reduces that overhead considerably as you just have one operation to monitor
+instead of multiple.
+
+**Q2: Why not model each device profile update as a firmware update that includes the desired software and configurations as well**
+
+Although this is a more robust approach, it is not feasible on all kinds of devices,
+especially the ones that does not support delta firmware updates.
+On such devices, pushing the entire firmware binary for each iterative change would considerably increase the binary size overhead.
+
+# Device profile capability
+
+A device that supports device profile operation must declare that capability by publishing the following message:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile' '{
+  "name": "prod-profile",
+  "version: "v1"
```

Comment:
From where are coming these name and version?

I guess from a previous successful device profile update. But what if meantime, *some* configuration files have been updated?

Reply:
Yes, the last applied device profile or the device profile with which the device was initially flashed in the factory. If such a factory profile is unavailable, that's when the agent can declare the empty JSON payload.

Reply:
> I guess from a previous successful device profile update. But what if meantime, some configuration files have been updated?

The question of "has anything changed since applying the device profile" is a tricky one, as I don't think there is one-way here, as there are multiple ways that the device profile state could get "out of sync" (e.g. sending individual firmware/software/config operations, user running manual commands via ssh etc.)

Though I don't think that device profiles are always used to control the whole on-device setup (as this usually leads to have an unwieldy amount of device profiles to manage, especially if a single device only needs 1 extra software package or configuration)

Below are some options (not exhaustive) that thin-edge.io could do:

**Option 1:** Do nothing

thin-edge.io doesn't do anything, and it is up to the device maintainer to control how the device is managed

**Option 2:** Flag profile as "dirty" after other operations

We introduce a "dirty" flag to the device profile current state (e.g. in the /twin setting), which is set anytime a single firmware/software/configuration operation is sent after a device profile is applied. The "dirty" flag would work similar to git, where it just indicates that something has changed (but does not say what).

There could also be a possible option made to even disable individual cloud operations, e.g. only accept cloud device profile operations, and reject cloud based firmware/software/configuration operations.

**Option 3:** Add interface which can be executed to check if the device profile is still 

Similar to the sm-plugin "list" command, a device profile handler could be used to indicate whether a device profile is still in the desired state (e.g. has not been affected by individual firmware/software/configuration operations). This would give a lot of control over the user deciding what is correct to do here (which could be nice).